### PR TITLE
[Frontend|Backend]Be able to troubleshoot a failed email/sms inject test

### PR DESCRIPTION
**Pull request by johanah29**

### CODE REVIEW

File: openbas-api/src/main/java/io/openbas/migration/V3_30__Add_Injects_tests_statuses.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+               status_name VARCHAR(255) NOT NULL,
+               status_executions text,
+               tracking_sent_date timestamp,
+               tracking_ack_date timestamp,
+               tracking_end_date timestamp,
+               tracking_total_execution_time bigint,
+               tracking_total_count int,
+               tracking_total_error int,
+               tracking_total_success int,
+               status_inject VARCHAR(255),
```
Comment (Dimfacion) [inline openbas-api/src/main/java/io/openbas/migration/V3_30__Add_Injects_tests_statuses.java:28]: There should be a foreign key here

File: openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/inject_test_status")
```
Comment (Dimfacion) [inline openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java:16]: I'm not a fan of this url. I think it's great that you switch for a dedicated controller but you should keep using the INJECT_URI from InjectApi.java

File: openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static io.openbas.config.SessionHelper.currentUser;
+
+@Service
+@Log
+@RequiredArgsConstructor
+public class InjectTestStatusService {
```
Comment (Dimfacion) [inline openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java:27]: It's nice to use a service here :+1:

File: openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    List<InjectTestStatus> injectTestStatuses = new ArrayList<>();
+    injects.forEach(inject -> {
+      if (injectTestStatusRepository.findByInject(inject).isPresent()) {
+        injectTestStatuses.add(injectTestStatusRepository.findByInject(inject).get());
+      }
+    });
+    return injectTestStatuses;
+  }
+
+  public List<InjectTestStatus> findAllScenarioInjectTests(String scenarioId) {
```
Comment (Dimfacion) [inline openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java:78]: Same as https://github.com/OpenBAS-Platform/openbas/pull/1311/files#r1711535703

File: openbas-front/src/admin/components/simulations/simulation/tests/ExerciseTestStatusDetail.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
+import { makeStyles } from '@mui/styles';
```
Comment (Dimfacion) [inline openbas-front/src/admin/components/simulations/simulation/tests/ExerciseTestStatusDetail.tsx:1]: Are there a lot of differences between this and ScenarioTestStatusDetail.tsx ? Can it be mutualized ?

File: openbas-api/src/main/java/io/openbas/rest/inject/InjectApi.java [hunk 6]
```diff
@@ -93,11 +88,7 @@ public class InjectApi extends RestBehavior {
   private final InjectService injectService;
   private final AtomicTestingService atomicTestingService;
   private final InjectDuplicateService injectDuplicateService;
-
-  @Autowired
-  public void setContext(ApplicationContext context) {
-    this.context = context;
-  }
+  private final InjectTestStatusService injectTestStatusService;
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/rest/inject/InjectApi.java:92]: This is not used

File: openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  public List<InjectTestStatus> findAllExerciseInjectTests(String exerciseId) {
+    return injectTestStatusRepository.findAllExerciseInjectTests(exerciseId);
+  }
+
+  public List<InjectTestStatus> findAllScenarioInjectTests(String scenarioId) {
+    return injectTestStatusRepository.findAllScenarioInjectTests(scenarioId);
+  }
+
+  public InjectTestStatus findInjectTestStatus(String testId) {
+    return injectTestStatusRepository.findById(testId).orElseThrow();
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java:76]: I'd prefer: findInjectTestStatusById

File: openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    this.injectTestStatusRepository.save(injectTestStatusToSave);
+
+    return injectTestStatusToSave;
+  }
+
+  public List<InjectTestStatus> findAllExerciseInjectTests(String exerciseId) {
+    return injectTestStatusRepository.findAllExerciseInjectTests(exerciseId);
+  }
+
+  public List<InjectTestStatus> findAllScenarioInjectTests(String scenarioId) {
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java:71]: I'd prefer findAllInjectTestsByScenarioId

File: openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      injectTestStatusToSave.setId(testStatus.getId());
+      injectTestStatusToSave.setTestCreationDate(testStatus.getTestCreationDate());
+    });
+    injectTestStatusToSave.setInject(inject);
+    this.injectTestStatusRepository.save(injectTestStatusToSave);
+
+    return injectTestStatusToSave;
+  }
+
+  public List<InjectTestStatus> findAllExerciseInjectTests(String exerciseId) {
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/service/InjectTestStatusService.java:67]: I'd prefer findAllInjectTestsByExerciseId

File: openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+@RestController
+@PreAuthorize("isAdmin()")
+@RequiredArgsConstructor
+public class InjectTestStatusApi extends RestBehavior {
+
+  private final InjectTestStatusService injectTestStatusService;
+
+  @GetMapping("/api/injects/{injectId}/test")
+  public InjectTestStatus testInject(@PathVariable String injectId) {
+    return injectTestStatusService.testInject(injectId);
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java:24]: Maybe you could add  @NotBlank  for injectId

File: openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  @GetMapping("/api/exercise/{exerciseId}/injects/test")
+  public List<InjectTestStatus> findAllExerciseInjectTests(@PathVariable String exerciseId) {
+    return injectTestStatusService.findAllExerciseInjectTests(exerciseId);
+  }
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java:30]: same

File: openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+  @GetMapping("/api/scenario/{scenarioId}/injects/test")
+  public List<InjectTestStatus> findAllScenarioInjectTests(@PathVariable String scenarioId) {
+    return injectTestStatusService.findAllScenarioInjectTests(scenarioId);
+  }
+
+  @GetMapping("/api/injects/test/{testId}")
+  public InjectTestStatus findInjectTestStatus(@PathVariable String testId) {
+    return injectTestStatusService.findInjectTestStatus(testId);
+  }
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java:40]: same

File: openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+  @JsonProperty("inject_title")
+  public String getInjectTitle() {
+    return inject.getTitle();
+  }
+
+  @JsonProperty("injector_contract")
+  public Optional<InjectorContract> getInjectContract() {
+    return inject.getInjectorContract();
+  }
```
Comment (savacano28) [inline openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java:90]: Do we need the whole contract?

File: openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+@Setter
+@Getter
+@Entity
+@Table(name = "injects_tests_statuses")
+public class InjectTestStatus implements Base {
+
```
Comment (savacano28) [inline openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java:26]: Could we implement a parent class status?

File: openbas-front/src/admin/components/common/injects/InjectPopover.tsx [hunk 15]
```diff
@@ -48,6 +68,8 @@ const InjectPopover: FunctionComponent<Props> = ({
   const [_injectTestResult, setInjectTestResult] = useState<InjectStatus | null>(null);
   const [anchorEl, setAnchorEl] = useState<Element | null>(null);
 
+  const isExercise = useHelper((helper: ExercisesHelper) => helper.getExercisesMap()[exerciseOrScenarioId!] === undefined);
+
```
Comment (savacano28) [inline openbas-front/src/admin/components/common/injects/InjectPopover.tsx:72]: Here you could make _different from undefined_ in order to know if the id is an exerciseId, or you could call this const: is NotExercise

File: openbas-front/src/admin/components/common/injects/InjectPopover.tsx [hunk 16]
```diff
@@ -114,9 +136,36 @@ const InjectPopover: FunctionComponent<Props> = ({
...
+    }
+  }, [openDialog]);
+
   const submitTest = () => {
     testInject(inject.inject_id).then((result: { data: InjectStatus }) => {
       setInjectTestResult(result.data);
+      setOpenDialog(true);
+      if (isExercise) {
+        setDetailsLink(`/admin/scenarios/${exerciseOrScenarioId}/tests/${result.data.status_id}`);
+      } else {
```
Comment (savacano28) [inline openbas-front/src/admin/components/common/injects/InjectPopover.tsx:166]: It is not consistent with the variable name

File: openbas-front/src/admin/components/scenarios/scenario/tests/ScenarioTests.tsx [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+  },
+};
+
+const ScenarioTests: FunctionComponent = () => {
+  // Standard hooks
+  const classes = useStyles();
+  const { t, fldt } = useFormatter();
+
+  const { scenarioId, statusId } = useParams() as { scenarioId: ScenarioStore['scenario_id'], statusId: InjectTestStatus['status_id'] };
+  const [selectedTest, setSelectedTest] = useState<InjectTestStatus | null>(null);
```
Comment (savacano28) [inline openbas-front/src/admin/components/scenarios/scenario/tests/ScenarioTests.tsx:61]: You can just make: scenario Is: string, and this file are the same as exercise test

File: openbas-front/src/admin/components/scenarios/scenario/Index.tsx [hunk 18]
```diff
@@ -162,6 +169,7 @@ const IndexScenarioComponent: FunctionComponent<{ scenario: ScenarioStore }> = (
               <Route path="" element={errorWrapper(Scenario)({ setOpenScenarioRecurringFormDialog })} />
               <Route path="definition" element={errorWrapper(ScenarioDefinition)()} />
               <Route path="injects" element={errorWrapper(Injects)()} />
+              <Route path="tests" element={errorWrapper(Tests)()} />
               {/* Not found */}
```
Comment (savacano28) [inline openbas-front/src/admin/components/scenarios/scenario/Index.tsx:173]: Maybe here we have to pass also statusId? : tests/:statusId?

File: openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java (hunk 13 above)
Comment (johanah29) [reply openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java:35]: Yes, it is necessary to use the InjectIcon component

File: openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+  @GetMapping("/api/exercise/{exerciseId}/injects/test")
+  public List<InjectTestStatus> findAllExerciseInjectTests(@PathVariable @NotBlank String exerciseId) {
+    return injectTestStatusService.findAllInjectTestsByExerciseId(exerciseId);
+  }
+
+  @GetMapping("/api/scenario/{scenarioId}/injects/test")
+  public List<InjectTestStatus> findAllScenarioInjectTests(@PathVariable String scenarioId) {
+    return injectTestStatusService.findAllInjectTestsByScenarioId(scenarioId);
```
Comment (savacano28) [inline openbas-api/src/main/java/io/openbas/rest/inject_test_status/InjectTestStatusApi.java:34]: Same

File: openbas-model/src/main/java/io/openbas/database/model/InjectStatus.java [hunk 20]
```diff
@@ -18,81 +18,12 @@
...
-  @JoinColumn(name = "status_inject")
-  @JsonIgnore
-  private Inject inject;
-
-  // region transient
-  public List<String> statusIdentifiers() {
-    return this.getTraces().stream().flatMap(ex -> ex.getIdentifiers().stream()).toList();
-  }
-  // endregion
+public class InjectStatus extends BaseInjectStatus implements Base {
```
Comment (savacano28) [inline openbas-model/src/main/java/io/openbas/database/model/InjectStatus.java:22]: BaseInjectStatus already implements the Base interface

File: openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+  @JsonProperty("inject_test_status_created_at")
+  private Instant testCreationDate;
+
+  @UpdateTimestamp
+  @Column(name = "status_updated_at")
+  @JsonProperty("inject_test_status_updated_at")
+  private Instant testUpdateDate;
+
+  public static InjectTestStatus fromExecutionTest(Execution execution) {
+    InjectTestStatus injectTestStatus = new InjectTestStatus();
```
Comment (savacano28) [inline openbas-model/src/main/java/io/openbas/database/model/InjectTestStatus.java:56]: You could include this method on you abstract class, and here you can make:  `public static InjectTestStatus fromExecutionTest(Execution execution) {
        InjectTestStatus injectTestStatus = new InjectTestStatus();
        injectTestStatus.**updateStatusFromExecution**(execution);
        return injectTestStatus;
    }`

**Review by savacano28 — APPROVED** [review]

**Pull request merged**